### PR TITLE
Fix explicit json null handling (#17788)

### LINKED ISSUE #17788: Explicit json null lost during unmarshalling
Unmarshalling json to a type containing a `JsonNode` field leads to misinterpretation of `JNull`.

### Example
```nim
import 
    std / [json, options]

type Foo = object
  a: Option[JsonNode]

let jsonFoo = parseJson("""{"a": null}""")
let foo = jsonFoo.to(Foo)

echo foo.a.get.kind

```

### Current Output
```
Error: unhandled exception: Can't obtain a value from a `none` [UnpackDefect]
```

### Expected Output
```
JNull
```

### Possible Solution

* The error could be in [json.nim](https://github.com/nim-lang/Nim/blob/devel/lib/pure/json.nim#L1150-L1156) line 1150-1156.
* If `jsonNode != nil` and `jsonNode.kind == JNull` and `T is JsonNode` , we should set `dst` to `some(newJNull())`.

### Additional Information
* In json, implicit and explicit null, i.e., not providing a value for a key and providing null for a key, have different semantics.
* In most use cases, these differences don't matter
* When handling with json patches the semantic difference matters, because explicitly setting a value to null is required
* This issue is blocking my project: I'm currently building a json-patch implementation for nim.

```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-04-19
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 0a10af5a2ce27638c8298e96866a779928122269
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by hnicke**
Resolves #17788 

Signed-off-by: Heiko Nickerl <mail@hnicke.de>

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -96,6 +96,11 @@
 - `json.%`,`json.to`, `jsonutils.formJson`,`jsonutils.toJson` now work with `uint|uint64`
   instead of raising (as in 1.4) or giving wrong results (as in 1.2).
 
+- `json.to[Foo]` now handles explicit nulls better. 
+  When type `Foo` which has a field `foo: Option[JsonNode]` was converted, 
+  a JNull value was converted to `Option[none]`.
```
Comment (konsumlamm) [inline changelog.md]:
```suggestion
  a JNull value was converted to `none(JsonNode)`. 
```

File: tests/stdlib/tjson.nim [hunk 2]
```diff
@@ -7,7 +7,7 @@ discard """
 Note: Macro tests are in tests/stdlib/tjsonmacro.nim
 ]#
 
-import std/[json,parsejson,strutils]
+import std/[json,parsejson,strutils, options]
```
Comment (konsumlamm) [inline tests/stdlib/tjson.nim]:
```suggestion
import std/[json, parsejson, strutils, options]
```

**Review by konsumlamm — CHANGES_REQUESTED** [review]

**Comment by timotheecour** [comment]
wait

**Review by timotheecour — CHANGES_REQUESTED** [review]
hold on please

**Comment by timotheecour** [comment]
## before PR:
* `std/json` is consistent with std/jsonutils (returning `none(JsonNode)` instead of `some(null)`)
* `Option[T]` doesn't special case `T` and is consistent with `Option[JsonNode]`:
* `Foo(a: none(JsonNode))` is invariant by `(%x).to(Foo)` (ie roundtrip preserved)
* `Foo(a: some(newJNull()))` is modified by `(%x).to(Foo)` (ie roundtrip not preserved)
```nim
when true:
  import std/json
  import std/options
  type
    Bar = ref object
      x: int
    Foo = object
      a: Option[Bar]
  proc `$`(a: Bar): string =
    if a == nil: "nil"
    else: $a.x
  block:
    let s = """{"a": null}"""
    echo s
    echo s.parseJson
    let foo = s.parseJson.to(Foo)
    echo foo
```
prints:
{"a": null}
{"a":null}
(a: none(Bar))

ie, returns a `none(Bar)`

## after PR:
* `std/json` is not consistent with std/jsonutils
* `Option[T]` behaves differently from `Option[JsonNode]`
* `Foo(a: none(JsonNode))` is modified by (%x).to(Foo)
* `Foo(a: some(newJNull()))` is invariant by (%x).to(Foo)

## conclusion
IMO behavior before this PR is preferable.
This is arguably an edge case, but the existing behavior is more consistent than after this PR (even if we also modified jsonutils).

`none(JsonNode)` is more common and useful than `some(newJNull())` (and more compact).

jsonutils has a hookable API, maybe you should use that instead (see fromJsonHook, toJsonHook) to customize fromJson/toJson the way you want. So instead we should IMO document the existing behavior, and possibly add a test showing how to get desired effect with fromJsonHook, toJsonHook

**Comment by hnicke** [comment]
> none(JsonNode) is more common and useful than some(newJNull()) (and more compact).

..though is has different semantics (implicit vs explicit null).

But I guess you are right. I will solve my problem with a custom json hook. Thank you.

**Pull request closed without merging**